drivers/can/can_mcp2515.c

```diff
@@ -442,6 +479,67 @@ static void mcp2515_tx_done(struct device *dev, u8_t tx_idx)
 	k_sem_give(&dev_data->tx_sem);
 }
 
+static enum can_state mcp2515_get_state(struct device *dev,
+					struct can_bus_err_cnt *err_cnt)
+{
+	u8_t eflg;
+	u8_t err_cnt_buf[2];
+	int ret;
+
+	ret = mcp2515_cmd_read_reg(dev, MCP2515_ADDR_EFLG, &eflg, sizeof(eflg));
+	if (ret < 0) {
+		LOG_ERR("Failed to read error register [%d]", ret);
+		return CAN_BUS_UNKNOWN;
+	}
+
+	if (err_cnt) {
+		ret = mcp2515_cmd_read_reg(dev, MCP2515_ADDR_TEC, err_cnt_buf,
+					   sizeof(err_cnt_buf));
+		if (ret < 0) {
+			LOG_ERR("Failed to read error counters [%d]", ret);
+			return CAN_BUS_UNKNOWN;
+		}
+
+		err_cnt->tx_err_cnt = err_cnt_buf[0];
+		err_cnt->rx_err_cnt = err_cnt_buf[1];
+	}
+
+	if (eflg & MCP2515_EFLG_TXBO) {
+		return CAN_BUS_OFF;
+	}
+
+	if ((eflg & MCP2515_EFLG_RXEP) || (eflg & MCP2515_EFLG_TXEP)) {
+		return CAN_ERROR_PASSIVE;
+	}
+
+	return CAN_ERROR_ACTIVE;
+}
+
+static void mcp2515_handle_errors(struct device *dev)
+{
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+	can_state_change_isr_t state_change_isr = dev_data->state_change_isr;
+	enum can_state state;
+	struct can_bus_err_cnt err_cnt;
+
+	state = mcp2515_get_state(dev, state_change_isr ? &err_cnt : NULL);
+
+	/* needs a lock because we are in a thread here */
```

Comment:
Similar situation as before with tx_mutex. Another thread calling mcp2515_send doesn't need to be blocked

Reply:
I was concerned about the `mcp2515_register_state_change_isr` since the function is backed up, I can remove the locks.